File: internal/stream/agentchat_test.go
// SentinelDesk
// A collaborative operating system for people and AI agents.
//
// Copyright 2026 Federico Pereira <fpereira@cnsoluciones.com>
//
// Licensed under the Apache License, Version 2.0.
//
// This product's name and logo are trademarks of Federico Pereira and are not
// covered by the license above. See the README for the trademark policy.
//
// SPDX-License-Identifier: Apache-2.0

package stream

// The conversation record: the half of ask_human that does NOT wait.
//
// What these cover is the difference between sending and displaying. A question
// answered in the agent's own terminal still has to reach the room so a chat
// panel can render it, and reaching the room must not be able to hurt the
// conversation it describes: it cannot block, it cannot fail the caller, and it
// cannot fall over on the one member guaranteed to be present — the agent,
// which has no Session at all. That last one is not hypothetical;
// delivery_test.go records the same nil dereference taking the whole daemon
// down once already.
//
// No WebRTC anywhere here, for the reason delivery_test.go gives: a Session
// with no data channel is a valid recipient that sends nothing, which is
// exactly the part under test. The send itself vanishes into a DataChannel a
// test cannot watch, so the assertions land on the two things that have gone
// wrong before and can be seen — which members were selected, and what the
// message says.

import (
	"encoding/json"
	"sync"
	"testing"
	"time"
)

// chatRoom builds a room holding some browser members and optionally the agent,
// without NewRoom, which would try to open the X display for peer pointers.
func chatRoom(humans int, withAgent bool) *Room {
	r := &Room{members: map[string]*roomMember{}, controller: ControlFree}
	for i := 1; i <= humans; i++ {
		id := string(rune('a' + i - 1))
		r.members[id] = &roomMember{id: id, session: &Session{}}
		r.order = append(r.order, id)
	}
	if withAgent {
		// The agent as the room really holds it: a member with no Session.
		r.members[agentID] = &roomMember{id: agentID, agent: true}
		r.order = append(r.order, agentID)
	}
	return r
}

// TestAgentChatSurvivesAMemberWithoutASession is the regression, stated first
// because it is the one that used to be fatal. The agent is always in the room
// when it is talking, and it never has a Session.
func TestAgentChatSurvivesAMemberWithoutASession(t *testing.T) {
	for _, tc := range []struct {
		name string
		room *Room
	}{
		{"agent alone", chatRoom(0, true)},
		{"agent and two browsers", chatRoom(2, true)},
		{"nobody at all", chatRoom(0, false)},
	} {
		t.Run(tc.name, func(t *testing.T) {
			// A panic fails this by crashing it, which is the outcome wanted: the
			// assertion is not on a value, it is that the daemon survives writing
			// its own transcript.
			tc.room.AgentChat(ChatTurn{ID: "ask-1", Role: "agent",
				Via: "console", Text: "which invoice did you mean?"})

			for i, s := range tc.room.receivers() {
				if s == nil {
					t.Fatalf("receiver %d is nil — this is the dereference that "+
						"took the daemon down", i)
				}
			}
		})
	}
}

// TestAgentChatReachesEveryLiveSession. A record that went only to the
// controller would show the conversation to one person and hide it from
// everyone else watching the same desktop — and the controller is frequently
// the agent, which cannot read it at all.
func TestAgentChatReachesEveryLiveSession(t *testing.T) {
	r := chatRoom(3, true)

	got := len(r.receivers())
	if got != 3 {
		t.Fatalf("the record would reach %d members, want 3 (every browser, "+
			"not the agent)", got)
	}
	r.AgentChat(ChatTurn{ID: "ask-7", Role: "human", Via: "console",
		Text: "the second one"})

	// And with nobody in a browser it reaches nobody, which is the correct
	// outcome rather than a failure: a transcript exists to be read.
	if got := len(chatRoom(0, true).receivers()); got != 0 {
		t.Errorf("%d receivers in a room holding only the agent, want 0", got)
	}
}

// TestAgentChatWireShape pins what a browser actually parses, by encoding
// through the production path rather than a copy of it.
//
// The type is "agent_chat" and NOT "question", which is the entire point of the
// change: a client that has never heard of this ignores an unknown `t` and
// draws nothing, whereas reusing "question" would have popped the dialog this
// exists to stop — one with nothing waiting on it, so nothing would ever close
// it either.
func TestAgentChatWireShape(t *testing.T) {
	got := decodeTurn(t, ChatTurn{
		ID: "ask-3", Role: "agent", Via: "console",
		Text: "which invoice did you mean?", Options: []string{"first", "second"},
	})

	if got["t"] != AgentChatType {
		t.Errorf("type is %v, want %q", got["t"], AgentChatType)
	}
	if got["t"] == "question" {
		t.Error("the record reuses the type the ask dialog listens for; every " +
			"existing client would draw a modal nothing can close")
	}
	for field, want := range map[string]any{
		"id": "ask-3", "role": "agent", "via": "console",
		"text": "which invoice did you mean?",
	} {
		if got[field] != want {
			t.Errorf("%s is %v, want %v", field, got[field], want)
		}
	}
	opts, _ := got["options"].([]any)
	if len(opts) != 2 || opts[0] != "first" || opts[1] != "second" {
		t.Errorf("options came out as %v", got["options"])
	}
	if at, _ := got["at"].(float64); at <= 0 {
		t.Errorf("no usable timestamp on the record: %v", got)
	}
}

// TestAgentChatOmitsWhatIsNotThere. A free-text answer has no buttons, and a
// client checking `msg.options.length` should not have to know the difference
// between an absent field and a null one.
func TestAgentChatOmitsWhatIsNotThere(t *testing.T) {
	got := decodeTurn(t, ChatTurn{ID: "ask-4", Role: "human", Text: "yes"})
	if raw, present := got["options"]; present {
		t.Errorf("options is present as %v for a free-text turn", raw)
	}
	if raw, present := got["via"]; present {
		t.Errorf("via is present as %v when unset", raw)
	}
}

// TestAgentChatCarriesTheSystemRole. A question that timed out has no human
// turn, and a panel showing the question alone would leave the reader waiting
// for an answer that is never coming — the silent failure this project ranks
// above a crash.
func TestAgentChatCarriesTheSystemRole(t *testing.T) {
	got := decodeTurn(t, ChatTurn{ID: "ask-5", Role: "system", Via: "room",
		Text: "nobody answered in 2m0s"})
	if got["role"] != "system" {
		t.Errorf("role is %v, want system", got["role"])
	}
	if got["text"] != "nobody answered in 2m0s" {
		t.Errorf("the reason did not survive: %v", got["text"])
	}
}

// TestAgentChatDoesNotBlock is the property that makes this safe to call from
// the middle of a conversation. Notice has it for the same reason: what is being
// described has already happened, and holding the agent until somebody
// acknowledges the description would punish the person for reading it.
func TestAgentChatDoesNotBlock(t *testing.T) {
	r := chatRoom(4, true)

	done := make(chan struct{})
	go func() {
		defer close(done)
		for i := 0; i < 500; i++ {
			r.AgentChat(ChatTurn{ID: "ask-1", Role: "agent", Text: "still here?"})
		}
	}()

	select {
	case <-done:
	case <-time.After(5 * time.Second):
		t.Fatal("AgentChat blocked; it is supposed to be fire and forget")
	}
}

// TestAgentChatIsSafeWhileTheRoomChanges. The record is emitted from the
// goroutine handling a tool call while browsers join and leave on their own.
// Reading the member list needs the lock — the defect this change also fixed in
// Notice, which was snapshotting without one. Run under -race to mean anything.
func TestAgentChatIsSafeWhileTheRoomChanges(t *testing.T) {
	r := chatRoom(2, true)

	stop := make(chan struct{})
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		for i := 0; ; i++ {
			select {
			case <-stop:
				return
			default:
			}
			id := string(rune('m' + i%8))
			r.mu.Lock()
			r.members[id] = &roomMember{id: id, session: &Session{}}
			r.mu.Unlock()
			r.mu.Lock()
			delete(r.members, id)
			r.mu.Unlock()
		}
	}()
	go func() {
		defer wg.Done()
		for i := 0; i < 1000; i++ {
			r.AgentChat(ChatTurn{ID: "ask-9", Role: "system",
				Text: "nobody answered"})
			r.Notice("credential", []string{"api key"})
		}
		close(stop)
	}()
	wg.Wait()
}

// --- what the record must NOT have changed ---------------------------------------

// TestAskStillRefusesAnEmptyRoomAtOnce.
//
// The record travels the same path as the question and it would have been easy
// to make asking succeed just because somebody could be told about it. Nobody in
// a browser is still nobody to ask, and it has to fail immediately rather than
// after the full timeout — the run that started all of this burned 120 seconds
// on a room where the refusal was the right answer from the first millisecond.
func TestAskStillRefusesAnEmptyRoomAtOnce(t *testing.T) {
	r := chatRoom(0, true)

	start := time.Now()
	answer, err := r.AskHuman("shall I carry on?", nil, 30*time.Second)
	took := time.Since(start)

	if err == nil {
		t.Fatalf("asking an empty room succeeded with %q", answer)
	}
	if took > 5*time.Second {
		t.Errorf("the refusal took %v; nobody being here is known immediately", took)
	}
}

// TestAskStillReportsATimeoutAsATimeout. A default returned here would make
// "nobody was looking" indistinguishable from "somebody chose this", which is
// the whole reason the tool exists.
func TestAskStillReportsATimeoutAsATimeout(t *testing.T) {
	// One browser member, which is alive as far as the room can tell — a Session
	// with no PeerConnection cannot be declared dead. That is precisely the
	// scenario behind the original defect: a tab open, nobody in front of it.
	r := chatRoom(1, true)

	answer, err := r.AskHuman("shall I delete it?", []string{"yes", "no"}, 50*time.Millisecond)
	if err == nil {
		t.Fatalf("an unanswered question came back as the answer %q", answer)
	}
	if answer != "" {
		t.Errorf("a timeout carried an answer: %q", answer)
	}
	// And the room is left ready for the next question rather than stuck on the
	// one that expired.
	r.mu.RLock()
	stuck := r.asking != nil
	r.mu.RUnlock()
	if stuck {
		t.Error("the expired question is still holding the room's single slot")
	}
}

// decodeTurn runs a turn through the real encoder and hands back what a browser
// would see.
func decodeTurn(t *testing.T, turn ChatTurn) map[string]any {
	t.Helper()
	raw, err := turn.encode()
	if err != nil {
		t.Fatalf("the record does not marshal: %v", err)
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("the record is not JSON: %v\n%s", err, raw)
	}
	return out
}

// TestTheHistoryFrameSaysWhichOfTheThreeItIs.
//
// agent_history carries one of three things and the panel has to tell them
// apart from the frame alone: the list of past conversations, one of them
// opened to read, and the conversation the runtime is IN — sent when somebody
// continues a past one, or when a terminal attaches and asks what is going on.
//
// The rule used to be the session number: 0 meant the list. That reading has no
// room for the third case, whose session IS 0, and it is why a continued
// conversation arrived as an empty history drawer instead of as a thread.
func TestTheHistoryFrameSaysWhichOfTheThreeItIs(t *testing.T) {
	list := agentHistoryPayload(0, "", []AgentSession{{ID: 24, Title: "a run"}}, nil)
	if _, ok := list["sessions"]; !ok {
		t.Error("the list of conversations arrived without a sessions field")
	}
	if _, ok := list["messages"]; ok {
		t.Error("the list of conversations arrived carrying messages")
	}

	past := agentHistoryPayload(24, "", nil, []AgentHistoryTurn{{Role: "human", Text: "hi"}})
	if _, ok := past["messages"]; !ok {
		t.Error("a transcript arrived without its messages")
	}
	if _, ok := past["chat"]; ok {
		t.Error("a transcript being READ named a chat id, which would move the panel into it")
	}

	live := agentHistoryPayload(0, "resume-24-99", nil, []AgentHistoryTurn{{Role: "human", Text: "hi"}})
	if _, ok := live["sessions"]; ok {
		t.Fatal("the LIVE conversation was framed as the sessions list — " +
			"the panel reads that as an empty history and keeps the old thread")
	}
	if live["chat"] != "resume-24-99" {
		t.Errorf("the live conversation arrived as chat %v, want resume-24-99", live["chat"])
	}
	if live["session"] != 0 {
		t.Errorf("the live conversation arrived as session %v, want 0", live["session"])
	}
}

// TestAnEmptyTranscriptStillCarriesTheField.
//
// The runtime's wire omits an empty array — so a terminal attaching to an idle
// engine gets a transcript with no messages field at all. This process is where
// that becomes an empty array again, and it has to: a panel replacing what it
// shows cannot tell a missing field from a frame that is not about the thread,
// and would read it as the sessions list and wipe the drawer.
func TestAnEmptyTranscriptStillCarriesTheField(t *testing.T) {
	got := agentHistoryPayload(0, "c-1", nil, []AgentHistoryTurn{})
	raw, ok := got["messages"]
	if !ok {
		t.Fatal("an empty transcript dropped its messages field")
	}
	if msgs, ok := raw.([]AgentHistoryTurn); !ok || msgs == nil {
		t.Errorf("messages came through as %#v, want an empty array", raw)
	}
	// And it survives the encoder as [] rather than null.
	var back map[string]json.RawMessage
	if err := json.Unmarshal([]byte(jsonLine(got)), &back); err != nil {
		t.Fatal(err)
	}
	if string(back["messages"]) != "[]" {
		t.Errorf("on the wire messages is %s, want []", back["messages"])
	}
}

// TestOnePersonInTwoWindowsIsOneHand.
//
// # What was on screen
//
// A room with one human in two browsers holds TWO members: same display name,
// different session ids, different colours from the rotation. The rule that
// hides the driver's marker — their hand is already on the X pointer, so a
// second arrow is a duplicate — matched only the session that held control.
// The other window went on drawing an arrow with the same name in another
// colour, shadowing the real pointer.
//
// The comparison is by NAME because that is the only identity a member has.
// The cost is stated where the function lives: two different people who share a
// display name are treated as one. That is the rarer case, and it was chosen.
func TestOnePersonInTwoWindowsIsOneHand(t *testing.T) {
	for _, tc := range []struct {
		driver, who string
		want        bool
		why         string
	}{
		{"Federico Pereira", "Federico Pereira", true, "the same person's other window"},
		{"Federico Pereira", " federico pereira ", true, "spacing and case are not identity"},
		{"Federico Pereira", "Ana Maria", false, "somebody else entirely"},
		{"", "Federico Pereira", false, "an unnamed driver must not suppress everyone"},
		{"Federico Pereira", "", false, "an unnamed viewer is not the driver"},
		{"", "", false, "two nothings are not the same person"},
		{"   ", "   ", false, "and neither are two blanks"},
	} {
		if got := sameHand(tc.driver, tc.who); got != tc.want {
			t.Errorf("sameHand(%q, %q) = %v, want %v — %s",
				tc.driver, tc.who, got, tc.want, tc.why)
		}
	}
}

File: app/src/transport/agentChat.test.ts
// SentinelDesk
// A collaborative operating system for people and AI agents.
//
// Copyright 2026 Federico Pereira <fpereira@cnsoluciones.com>
//
// Licensed under the Apache License, Version 2.0.
//
// This product's name and logo are trademarks of Federico Pereira and are not
// covered by the license above. See the README for the trademark policy.
//
// SPDX-License-Identifier: Apache-2.0

/* The chat panel's pure logic.
 *
 * This file exists because of two bugs that shipped, and both of them were the
 * kind a test here catches in a second: an answer drawn twice because one wire
 * field has two names, and a row of `{{n}}` because a placeholder was written in
 * the wrong brace style. Neither needed a browser, a desktop or an agent to
 * reproduce — only the functions below, called directly.
 *
 * What is tested here is deliberately the part with no React in it. The panel's
 * rendering is verified by looking at it; what cannot be verified by looking is
 * whether a slash is a command, and that is exactly what this covers. */

import { describe, expect, it } from 'vitest'

import {
  classifyHistory,
  isRunning,
  matchCommand,
  restoreTurns,
  settleStreaming,
  type AgentCommand,
  type AgentMessage,
} from './agentChat'

/* The surface as an engine announces it. Written out rather than imported from
 * the panel, because there is nothing in the panel to import any more — the
 * list lives in the runtime, and that is the property being relied on. */
const OFFERED: readonly AgentCommand[] = [
  { name: '/compact', kind: 'compact', id: 'cmd.compact', what: 'fold the conversation' },
  { name: '/model', kind: 'model', id: 'cmd.model', what: 'switch model' },
  { name: '/stop', kind: '', id: 'cmd.stop', what: 'stop the agent' },
]

describe('matchCommand', () => {
  it('recognises a command the engine announced', () => {
    expect(matchCommand('/compact', OFFERED)).toEqual({ kind: 'compact', text: '' })
  })

  it('carries the argument', () => {
    expect(matchCommand('/model claude-code/opus', OFFERED)).toEqual({
      kind: 'model',
      text: 'claude-code/opus',
    })
  })

  it('keeps an argument that has spaces in it', () => {
    expect(matchCommand('/model  a  b ', OFFERED)?.text).toBe('a  b')
  })

  /* The rule that goes the opposite way from the terminal's, on purpose.
   *
   * A terminal prompt is a command line where a slash is a verb, so an unknown
   * one is a mistake and is refused. This is a chat box, where somebody may
   * reasonably begin a sentence with a path — refusing "/opt/desktop is where
   * it lives" would be the panel inventing a syntax error in the middle of a
   * conversation. */
  it('treats an unknown slash as a message, not an error', () => {
    expect(matchCommand('/opt/desktop is where it lives', OFFERED)).toBeNull()
    expect(matchCommand('/nonsense', OFFERED)).toBeNull()
  })

  it('is not fooled by a slash inside a sentence', () => {
    expect(matchCommand('use and/or', OFFERED)).toBeNull()
    expect(matchCommand('what about /compact', OFFERED)).toBeNull()
  })

  /* Nothing is offered until the engine has said what it has. A palette that
   * listed commands before anything announced them would be making them up —
   * and would send verbs to a runtime that may never have had them. */
  it('offers nothing before the engine has announced anything', () => {
    expect(matchCommand('/compact', [])).toBeNull()
  })

  /* An empty kind means the PANEL handles it — /stop is a cancel, not a
   * command — and the caller has to be able to tell that from "not a command"
   * without guessing. */
  it('distinguishes a panel-handled command from an unknown one', () => {
    expect(matchCommand('/stop', OFFERED)).toEqual({ kind: '', text: '' })
    expect(matchCommand('/unknown', OFFERED)).toBeNull()
  })
})

/** classifyHistory is the rule that decides what an agent_history frame IS.
 *
 * The bug it exists to prevent had no symptom a stack trace could show: a
 * transcript of the live conversation — the frame the runtime sends when
 * somebody continues a past chat, or when a terminal attaches and asks what is
 * going on — carries session 0, and the panel used to read every session-0
 * frame as the sessions list. The restored conversation went into the history
 * drawer as an empty list, and the thread stayed as it was.
 */
describe('classifyHistory', () => {
  it('reads a frame with no messages as the sessions list', () => {
    expect(classifyHistory({ session: 0, sessions: [] })).toBe('list')
  })

  it('reads session 0 WITH messages as the live conversation', () => {
    expect(classifyHistory({ session: 0, messages: [{ role: 'human', text: 'hi' }] })).toBe('live')
  })

  it('reads a numbered session with messages as one to look at', () => {
    expect(classifyHistory({ session: 24, messages: [] })).toBe('past')
  })

  it('does not mistake an empty live conversation for the list', () => {
    // The runtime answers a transcript request with an empty array when there
    // is nothing to show, and that is still a transcript. Reading it as the
    // list would wipe the history drawer every time a terminal attached to an
    // idle engine.
    expect(classifyHistory({ session: 0, messages: [] })).toBe('live')
  })
})

/** isRunning decides whether an incoming frame joins the bubble in flight.
 *
 * The bug: both sides spell "nothing" as the empty string and mean opposite
 * things by it. Nothing is running is busy === ''. This frame belongs to no
 * conversation is chat === ''. Compared directly they matched, so a note that
 * arrived outside any conversation — emptying the history, with no chat open —
 * opened a streaming bubble. Nothing closes a streaming bubble except the end
 * of an exchange, and there was no exchange: "Thinking…" stayed on screen for
 * ever with "forgot 15 conversation(s)" inside it.
 */
describe('isRunning', () => {
  it('does not read a frame with no conversation as the running one', () => {
    expect(isRunning('', '')).toBe(false)
  })

  it('is true only for the exchange actually in flight', () => {
    expect(isRunning('c-5', 'c-5')).toBe(true)
    expect(isRunning('c-5', 'c-4')).toBe(false)
    expect(isRunning('c-5', '')).toBe(false)
  })

  it('does not read a conversationless frame as running even when one is', () => {
    expect(isRunning('', 'c-5')).toBe(false)
  })
})

/** settleStreaming closes bubbles the runtime can no longer finish.
 *
 * Clearing `busy` gave the composer back and left the spinner turning above it.
 * A streaming bubble is a promise that more text is coming; when the runtime
 * disconnects, nothing will ever send it, and the panel goes on claiming an
 * answer is on its way for the rest of the session.
 */
describe('settleStreaming', () => {
  const row = (streaming: boolean): AgentMessage => ({
    key: 'k' + String(streaming),
    chat: 'c-1',
    role: 'agent',
    text: '',
    at: 0,
    streaming,
    steps: [],
    ending: null,
    streamed: streaming,
  })

  it('closes an open bubble', () => {
    const out = settleStreaming([row(true)])
    expect(out[0].streaming).toBe(false)
  })

  it('returns the same array when nothing is open, so no render is caused', () => {
    const quiet = [row(false)]
    expect(settleStreaming(quiet)).toBe(quiet)
  })

  it('leaves the text and the steps alone — only the promise is withdrawn', () => {
    const mid: AgentMessage = { ...row(true), text: 'half an ans' }
    const out = settleStreaming([mid])
    expect(out[0].text).toBe('half an ans')
    expect(out[0].streamed).toBe(true)
  })
})

/** restoreTurns is the shape a reloaded conversation comes back in.
 *
 * The bug it fixes was purely visual and therefore invisible to every layer
 * that could have caught it: the record was right, the wire was right, the
 * panel drew each turn faithfully — as its own bubble, with its own fold. A
 * twelve-turn run that had been ONE chain of tool calls while it ran came back
 * from history as twelve headers reading "Tools · 2".
 *
 * So what is asserted here is the property the eye was checking: consecutive
 * agent turns become one bubble, carrying every step in order, and a human
 * turn still stands on its own.
 */
describe('restoreTurns', () => {
  let n = 0
  const mkKey = () => `k${++n}`
  const turn = (role: string, text: string, tools: string[] = []) => ({
    role,
    text,
    steps: tools.map((tool) => ({ tool, detail: '' })),
  })

  it('folds a run into one bubble however many turns it took', () => {
    const out = restoreTurns(
      [
        turn('human', 'play the video'),
        turn('agent', 'reading the skill', ['skill_read']),
        turn('agent', 'taking control', ['request_control']),
        turn('agent', 'opening it', ['browser_open']),
      ],
      'past-79',
      mkKey,
    )
    expect(out).toHaveLength(2)
    expect(out[0].role).toBe('human')
    expect(out[1].steps.map((s) => s.tool)).toEqual([
      'skill_read',
      'request_control',
      'browser_open',
    ])
  })

  it('reads as paragraphs, not as one run-on sentence', () => {
    const out = restoreTurns([turn('agent', 'first'), turn('agent', 'second')], 'c', mkKey)
    expect(out[0].text).toBe('first\n\nsecond')
  })

  it('does not add an empty paragraph for a turn that only called tools', () => {
    const out = restoreTurns(
      [turn('agent', 'said something'), turn('agent', '', ['browser_wait_until'])],
      'c',
      mkKey,
    )
    expect(out[0].text).toBe('said something')
    expect(out[0].steps).toHaveLength(1)
  })

  it('starts a new bubble after a human turn, so a continued conversation is not one blob', () => {
    const out = restoreTurns(
      [turn('agent', 'done'), turn('human', 'now this'), turn('agent', 'on it')],
      'c',
      mkKey,
    )
    expect(out.map((m) => m.role)).toEqual(['agent', 'human', 'agent'])
  })

  it('gives every bubble a key of its own', () => {
    const out = restoreTurns([turn('human', 'a'), turn('agent', 'b')], 'c', mkKey)
    expect(new Set(out.map((m) => m.key)).size).toBe(out.length)
  })

  it('restores nothing as nothing rather than as an empty bubble', () => {
    expect(restoreTurns([], 'c', mkKey)).toEqual([])
  })
})
